tests/Functional/Driver/PDO/MySQL/DeadlockTest.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace Doctrine\DBAL\Tests\Functional\Driver\PDO\MySQL;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver as DriverInterface;
+use Doctrine\DBAL\Driver\PDO\MySQL\Driver;
+use Doctrine\DBAL\Exception\DeadlockException;
+use Doctrine\DBAL\Tests\Functional\Driver\AbstractDriverTest;
+use Doctrine\DBAL\Tests\TestUtil;
+use ReflectionObject;
+use Throwable;
+
+use function pcntl_fork;
+use function sleep;
+
+class DeadlockTest extends AbstractDriverTest
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (TestUtil::isDriverOneOf('pdo_mysql')) {
+            return;
+        }
+
+        self::markTestSkipped('This test requires the pdo_mysql driver.');
+    }
+
+    public function testNestedTransactionsDeadlockExceptionHandling(): void
+    {
+        $firstConnection = new Connection($this->connection->getParams(), $this->driver);
+        $firstConnection->setNestTransactionsWithSavepoints(true);
+        $secondConnection = new Connection($this->connection->getParams(), $this->driver);
+        $secondConnection->setNestTransactionsWithSavepoints(true);
+
+        $this->assertNotSame($firstConnection, $secondConnection);
+
+        $this->prepareDatabase($firstConnection);
+
+        $executeQueries = static function (Connection $conn, int $thread): void {
+            $conn->beginTransaction();
+            $conn->beginTransaction();
+            if ($thread === 1) {
+                $conn->executeQuery('DELETE FROM `test1`');
+                sleep(2);
+                $conn->executeQuery('DELETE FROM `test2`');
+            } else {
+                $conn->executeQuery('DELETE FROM `test2`');
+                $conn->executeQuery('DELETE FROM `test1`');
+            }
+
+            $conn->commit();
+            $conn->commit();
+        };
+
+        $pid = pcntl_fork();
```

Comment:
Why does the test need multiple processes? Couldn't the single process use both connections?

Reply:
I've tried to do it with a single process, but as long as the second connection makes the deletion from test2 it gets into a table lock state and waits to be resolved, thus blocking the thread.

So this is the only way I came across to deal with it.

Reply:
This makes sense. Thanks for the details.